bakefile-0.2.9/doc/man/bakefile.1
'\" t
.\"     Title: bakefile
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 02/02/2011
.\"    Manual: Running Bakefile
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "BAKEFILE" "1" "02/02/2011" "[FIXME: source]" "Running Bakefile"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
bakefile \- native makefiles generator
.SH "SYNOPSIS"
.HP \w'\fBbakefile\fR\ 'u
\fBbakefile\fR [\fB\-\-version\fR] [\fB\-\-help\fR] \fB\-f\fR\fIFORMAT\fR \fB\-o\fR\fIOUTFILE\fR [\fB\-\-eol=[format|dos|unix|mac|native]\fR] [\fB\-\-wrap\-output=[no|\fR\fB\fILENGTH\fR\fR\fB]\fR] [\fB\-D\fR\fIVAR\fR=\fIVALUE\fR...] [\fB\-I\fR\fIPATH\fR...] [\fB\-v\fR] [\fB\-q\fR] [\fB\-\-dry\-run\fR] [\fB\-\-touch\fR] [\fB\-\-dump\fR] \fIfile\&.bkl\fR
.SH "DESCRIPTION"
.PP

\fBbakefile\fR
creates various types of Makefiles and project files from a single project description called a "Bakefile"\&.
.SH "COMMAND LINE OPTIONS"
.PP
.PP
\fB\-\-version\fR
.RS 4
Display Bakefile version and exit\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display usage information and exit\&.
.RE
.PP
\fB\-f\fR\fIFORMAT\fR, \fB\-\-format\fR=\fIFORMAT\fR
.RS 4
Specify output format\&. Bakefile supports the following formats:
.sp
.TS
allbox tab(:);
lB lB.
T{
Format
T}:T{
File(s) Generated
T}
.T&
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l.
T{
autoconf
T}:T{
Makefile\&.in for GNU Autoconf
T}
T{
borland
T}:T{
Makefile for Borland C++ and Borland
                        make
T}
T{
dmars
T}:T{
Generic Makefile for Digital Mars C/C++
T}
T{
dmars_smake
T}:T{
Makefile for Digital Mars C/C++ with
                        SMAKE
T}
T{
gnu
T}:T{
Makefile for GNU toolchain: GNU Make, GCC,
                        etc\&.
T}
T{
mingw
T}:T{
Makefile for MinGW toolchain:
                        mingw32\-make, MinGW port of GCC, etc\&.
T}
T{
msvc
T}:T{
Makefile for Visual C++ with Microsoft
                        nmake
T}
T{
msvc6prj
T}:T{
Microsoft Visual C++ 6\&.0 project files
T}
T{
msevc4prj
T}:T{
Microsoft Embedded Visual C++ 4 project
                        files
T}
T{
msvs2003prj
T}:T{
MS Visual Studio 2003 project files
T}
T{
msvs2005prj
T}:T{
MS Visual Studio 2005 project files
T}
T{
msvs2008prj
T}:T{
MS Visual Studio 2008 project files
T}
T{
suncc
T}:T{
GNU makefile for SunCC compiler
T}
T{
symbian
T}:T{
Symbian development files
T}
T{
watcom
T}:T{
Makefile for OpenWatcom C/C++
T}
T{
xcode2
T}:T{
Apple Xcode 2\&.4 project files
T}
.TE
.sp 1
.RE
.PP
\fB\-o\fR\fIOUTFILE\fR, \fB\-\-output\fR=\fIOUTFILE\fR
.RS 4
File to write generated makefile to\&. For those backends that generate more than one file, this option specifies the name of the main makefile\&.
.sp
This option has special meaning for msvs200xprj formats: by default, both the project files (one for each target) and a solution file, containing all the project files, are generated\&. However if OUTFILE is a file with \&.vcproj extension, then only the (necessarily unique) project file will be generated\&.
.RE
.PP
\fB\-D\fR\fIVAR\fR=\fIVALUE\fR
.RS 4
Define Bakefile variable\&. This definition overrides any definition from the ruleset or input makefile\&. You can use it to customize generated output\&.
.RE
.PP
\fB\-\-eol\fR=\fI[format|dos|unix|mac|native]\fR
.RS 4
Change the type of line endings used by general files\&.
dos,
unix
and
mac
specify the line endings used by respective platforms\&.
native
will use line endings of the platform Bakefile is ran on (doing this is usually a bad idea, but it is useful e\&.g\&. when checking generated files into RCS system that can\*(Aqt deal with line endings correctly, such as CVS)\&. The default value is
format
and means that the most appropriate line endings for the output format will be used \- Windows makefiles will use DOS line endings, Autoconf makefiles will use Unix ones and so on\&.
.RE
.PP
\fB\-\-wrap\-output\fR=\fI[no|LENGTH]\fR
.RS 4
Change line wrappings behavior\&. By default, Bakefile wraps generated makefiles so that lines don\*(Aqt exceed 75 characters\&. Use this option to either change the limit or to disable wrapping entirely by using
no
as the value\&.
.RE
.PP
\fB\-I\fR\fIPATH\fR
.RS 4
Add path to the list of directories where Bakefile looks for rules and output templates\&.
.RE
.PP
\fB\-\-dry\-run\fR
.RS 4
Process the bakefile normally, but instead of creating or modifying files, just print which files would be changed without actually modifying them\&.
.RE
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Be verbose\&.
.RE
.PP
\fB\-q\fR, \fB\-\-quiet\fR
.RS 4
Supress all output except for errors\&.
.RE
.PP
\fB\-\-touch\fR
.RS 4
Always touch output files, even if their content doesn\*(Aqt change\&.
.RE
.PP
\fB\-\-debug\fR
.RS 4
show internal debugging information
.RE
.PP
\fB\-\-dump\fR
.RS 4
Dump all Bakefile variables and targets to standard output instead of generating output\&. This is only useful for debugging Bakefile or ill\-behaving makefiles\&.
.RE
.PP
\fB\-\-output\-deps\fR=\fIFILE\fR
.RS 4
Output dependency information needed by
\fBbakefile_gen\fR
utility
.RE
.PP
\fB\-\-output\-changes\fR=\fIFILE\fR
.RS 4
Store list of changed files to the given file
.RE
.PP
\fB\-\-xml\-cache\fR=\fIFILE\fR
.RS 4
specify cache file where
\fBbakefile_gen\fR
stores pre\-parsed XML files
.RE
.SH "ENVIRONMENT VARIABLES"
.PP
.PP
BAKEFILE_PATHS
.RS 4
List of directories where ruleset files are looked for (syntax is same as in
\fIPATH\fR)\&. Bakefile\*(Aqs data directory is always searched in addition to paths listed in
\fIBAKEFILE_PATHS\fR, but
\fIBAKEFILE_PATHS\fR
has higher priority\&.
.RE


bakefile-0.2.9/src/bakefile.py
#!/usr/bin/env python

#
#  This file is part of Bakefile (http://www.bakefile.org)
#
#  Copyright (C) 2003-2007 Vaclav Slavik
#
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
#  IN THE SOFTWARE.
#
#  $Id: bakefile.py 1296 2009-03-17 05:14:32Z vaclavslavik $
#

import sys, os.path
from optparse import OptionParser

import errors
import formats
import xmlparser

def addIncludePaths(includes):
    import config
    normalized = [os.path.normpath(p) for p in includes]
    config.searchPath = normalized + config.searchPath

class BakefileOptionParser(OptionParser):
    def __init__(self):
        from version import BAKEFILE_VERSION
        OptionParser.__init__(self,
                              version='Bakefile %s' % BAKEFILE_VERSION,
                              usage='usage: %prog [options] inputfile.bkl')
    def print_help(self, file=None):
        # prepare list of available formats; we have to do it here, because
        # showFormats() may print warnings or errors to stderr and we don't
        # want that to happen in the middle of printing help
        if self.values.includes != None:
            addIncludePaths(self.values.includes)
        formatsList = formats.showFormats()

        if file is None:
            file = sys.stdout
        OptionParser.print_help(self, file)
   
        # show all available formats:
        file.write('\n%s' % formatsList)


def run(args):
    import config

    if sys.version_info < (2,5):
        sys.path.append(os.path.join(config.progdir, 'py25modules'))

    parser = BakefileOptionParser()
    parser.add_option('-f', '--format',
                      action="store", dest='format',
                      help='format of generated makefile')
    parser.add_option('-o', '--output',
                      action="store", dest='outfile',
                      help='output file')
    parser.add_option('-D',
                      action="append", dest='defines', metavar='VAR=VALUE',
                      help='override variable or option definition')
    parser.add_option('-I',
                      action="append", dest='includes', metavar='PATH',
                      help='search for bakefile rules in PATH')
    parser.add_option('-v', '--verbose',
                      action="store_true", dest='verbose', default=0,
                      help='display detailed information')
    parser.add_option('-q', '--quiet',
                      action="store_true", dest='quiet', default=0,
                      help='supress all output except of errors')
    parser.add_option('', '--dry-run',
                      action="store_true", dest='dry_run', default=0,
                      help="don't write any files, just pretend doing it")
    parser.add_option('', '--touch',
                      action="store_true", dest='always_touch_output', default=0,
                      help="always touch output files, even if their content doesn't change")
    parser.add_option('', '--eol',
                      default="format", action="store", dest='eol',
                      metavar='STYLE', type='choice',
                      choices=['format','dos','unix','mac','native'],
                      help="line endings type to use in output files (format, dos, unix, mac, native) [default: format]")
    parser.add_option('', '--wrap-output',
                      default="75", action="store", dest='wrap_lines_at',
                      metavar='LENGTH',
                      help="don't generate lines longer than LENGTH; set to \"no\" to disable wrapping [default: 75]")
    parser.add_option('', '--output-deps',
                      action="store", dest='deps_file', metavar='DEPSFILE',
                      help="output dependencies information for bakefile_gen")
    parser.add_option('', '--output-changes',
                      action="store", dest='changes_file', metavar='MODSFILE',
                      help="output list of modified files to a file")
    parser.add_option('', '--xml-cache',
                      action="store", dest='xml_cache', metavar='CACHEFILE',
                      help="cache file where bakefile_gen stores pre-parsed XML files")
    parser.add_option('', '--debug',
                      action="store_true", dest='debug', default=0,
                      help="show debugging information (you don't want this)")
    parser.add_option('', '--dump',
                      action="store_true", dest='dump', default=0,
                      help='dump parsed makefile content instead of '+
                           'generating makefile')

    options, args = parser.parse_args(args)

    if len(args) != 1:
        parser.error('incorrect number of arguments, exactly 1 .bkl required')

    config.dry_run = options.dry_run
    config.always_touch_output = options.always_touch_output
    config.debug = options.debug
    config.quiet = options.quiet
    if config.quiet:
        config.verbose = 0
    else:
        config.verbose = options.verbose

    if options.includes != None:
        addIncludePaths(options.includes)

    if options.xml_cache != None:
        try:
            # don't use shelve.open(), limit ourselves to dbhash version by
            # using the Shelf class explicitly; that's because we store large
            # chunks of data in the shelve and some dbm implementations can't
            # handle it -- see http://www.bakefile.org/ticket/214
            import pickle, shelve, dbhash
            xmlparser.cache = shelve.Shelf(dbhash.open(options.xml_cache, 'c'),
                                           protocol=pickle.HIGHEST_PROTOCOL)
        except ImportError:
            sys.stderr.write("Warning: disabling XML cache because it's not supported by this version of Python\n")
        except AttributeError: # python < 2.3 didn't have HIGHEST_PROTOCOL
            sys.stderr.write("Warning: disabling XML cache because it's not supported by Python 2.2\n")
        except KeyError: # python < 2.3 didn't have protocol argument
            sys.stderr.write("Warning: disabling XML cache because it's not supported by Python 2.2\n")
        except Exception, e:
            sys.stderr.write("Warning: disabling XML cache because an error occured while loading %s:\n         %s\n" % (options.xml_cache, e))

    formats.loadFormats()
    if options.format == None:
        parser.error('you must specify output format (use -f option)')
    config.format = options.format
    if not formats.isValidFormat(config.format):
        parser.error("unknown format '%s'\n\n" % config.format +
                     formats.showFormats())

    if options.deps_file != None:
        config.track_deps = 1
        config.deps_file = options.deps_file

    if options.changes_file != None:
        config.changes_file = options.changes_file
    
    if options.outfile != None:
        config.output_file = options.outfile
    else:
        fmt = formats.formats[config.format]
        if fmt.defaultFile == None or fmt.defaultFile == '':
            parser.error('you must specify output file (use -o option)')
        else:
            config.output_file = \
                os.path.join(os.path.dirname(args[0]), fmt.defaultFile)

    config.eol = options.eol

    if options.wrap_lines_at == "no":
        config.wrap_lines_at = None
    else:
        try:
            config.wrap_lines_at = int(options.wrap_lines_at)
        except ValueError:
            parser.error('invalid --wrap-output value: must be an integer or "no"')

    
    config.defines = {}
    if options.defines != None:
        for define in options.defines:
            d = define.split('=')
            if len(d) == 1:
                config.defines[d[0]] = ''
            else:
                config.defines[d[0]] = '='.join(d[1:])

    import reader, writer
    
    try:
        read_ok = reader.read(args[0])
    finally:
        if xmlparser.cache != None:
            xmlparser.cache.close()
            xmlparser.cache = None

    if not read_ok:
        sys.exit(1)

    if options.dump:
        import mk_dump
        mk_dump.dumpMakefile()
    else: 
        if not writer.write():
            sys.exit(1)

    if config.track_deps:
        import dependencies
        dependencies.save(config.deps_file)

if __name__ == '__main__':
    if sys.version_info[0:3] < (2,3,0):
        sys.stderr.write('error: Bakefile requires at least Python 2.3.0\n')
        sys.exit(1)

    do_profiling = 0 # set to 1 if debugging bottlenecks
    
    try:
        if do_profiling:
            import hotshot
            prof = hotshot.Profile('bakefile.prof')
            prof.runcall(run, sys.argv[1:])
            prof.close()
        else:
            run(sys.argv[1:])    
    except errors.ErrorBase, e:
        sys.stderr.write('%s\n' % e)
        sys.exit(1)
    except KeyboardInterrupt:
        sys.stderr.write('\nerror: bakefile cancelled by user\n')
        sys.exit(2)
